File: src/lib/entropy.ts
/**
 * Entropy estimation and crack-time modelling.
 * Mirrors Password-Strength-Analyzer/entropy.py so the preview and the
 * Flask backend always agree on the numbers they report.
 */

/** Guesses per second assumed for an offline attacker with modern GPUs. */
export const GUESSES_PER_SECOND = 1e10;

export type EntropyLabel =
  | "Very Weak"
  | "Weak"
  | "Reasonable"
  | "Strong"
  | "Very Strong";

/** Size of the character pool the password draws from. */
export function charsetSize(password: string): number {
  let size = 0;
  if (/[a-z]/.test(password)) size += 26;
  if (/[A-Z]/.test(password)) size += 26;
  if (/[0-9]/.test(password)) size += 10;
  if (/[ !-/:-@[-`{-~]/.test(password)) size += 33;
  // Anything outside printable ASCII (emoji, accents, ...).
  if (/[^\x20-\x7e]/.test(password)) size += 100;
  return size;
}

/** Shannon-style entropy: log2(charset ^ length), rounded to 1 decimal. */
export function calculateEntropy(password: string): number {
  const size = charsetSize(password);
  if (!size || !password.length) return 0;
  return Math.round(password.length * Math.log2(size) * 10) / 10;
}

/** Qualitative label for a given entropy value. */
export function entropyLabel(bits: number): EntropyLabel {
  if (bits < 28) return "Very Weak";
  if (bits < 36) return "Weak";
  if (bits < 60) return "Reasonable";
  if (bits < 128) return "Strong";
  return "Very Strong";
}

const TIME_UNITS: [number, string][] = [
  [60, "second"],
  [60, "minute"],
  [24, "hour"],
  [365, "day"],
  [100, "year"],
];

/** Human readable crack time, e.g. "142 years" or "instantly". */
export function estimateCrackTime(bits: number): string {
  if (bits <= 0) return "instantly";

  // Average attacker needs half the keyspace.
  const seconds = Math.pow(2, Math.min(bits, 1024)) / 2 / GUESSES_PER_SECOND;

  if (seconds < 1) return "instantly";

  let value = seconds;
  let unit = "second";
  for (const [factor, nextUnit] of TIME_UNITS) {
    if (value < factor) break;
    value /= factor;
    unit = nextUnit;
  }

  if (unit === "year" && value >= 100) {
    const centuries = value / 100;
    if (centuries >= 1e6) return "millions of centuries";
    return `${formatNumber(centuries)} centuries`;
  }

  const rounded = value < 10 ? Math.round(value * 10) / 10 : Math.round(value);
  return `${formatNumber(rounded)} ${unit}${rounded === 1 ? "" : "s"}`;
}

function formatNumber(value: number): string {
  if (value >= 1e6) return value.toExponential(2).replace("e+", " x 10^");
  return value.toLocaleString("en-US", { maximumFractionDigits: 1 });
}

File: Password-Strength-Analyzer/entropy.py
"""Entropy estimation and crack-time modelling."""

from __future__ import annotations

import math
import re

# Guesses per second assumed for an offline attacker with modern GPUs.
GUESSES_PER_SECOND = 1e10

# (threshold_in_bits, label) evaluated in order.
ENTROPY_LABELS: list[tuple[float, str]] = [
    (28, "Very Weak"),
    (36, "Weak"),
    (60, "Reasonable"),
    (128, "Strong"),
    (math.inf, "Very Strong"),
]

# (divisor, next_unit_name)
_TIME_UNITS: list[tuple[int, str]] = [
    (60, "minute"),
    (60, "hour"),
    (24, "day"),
    (365, "year"),
    (100, "century"),
]


def charset_size(password: str) -> int:
    """Return the size of the character pool the password draws from."""
    size = 0
    if re.search(r"[a-z]", password):
        size += 26
    if re.search(r"[A-Z]", password):
        size += 26
    if re.search(r"[0-9]", password):
        size += 10
    if re.search(r"[ !-/:-@\[-`{-~]", password):
        size += 33
    if re.search(r"[^\x20-\x7e]", password):  # emoji, accents, etc.
        size += 100
    return size


def calculate_entropy(password: str) -> float:
    """Return entropy in bits: length * log2(charset size)."""
    size = charset_size(password)
    if not size or not password:
        return 0.0
    return round(len(password) * math.log2(size), 1)


def entropy_label(bits: float) -> str:
    """Return the qualitative label for an entropy value."""
    for threshold, label in ENTROPY_LABELS:
        if bits < threshold:
            return label
    return "Very Strong"


def _format_number(value: float) -> str:
    if value >= 1_000_000:
        return f"{value:.2e}".replace("e+", " x 10^")
    return f"{value:,.1f}".rstrip("0").rstrip(".")


def estimate_crack_time(bits: float) -> str:
    """Return a human readable crack time such as '142 years'."""
    if bits <= 0:
        return "instantly"

    # An average attacker needs to search half the keyspace.
    seconds = (2 ** min(bits, 1024)) / 2 / GUESSES_PER_SECOND
    if seconds < 1:
        return "instantly"

    value, unit = seconds, "second"
    for divisor, next_unit in _TIME_UNITS:
        if value < divisor:
            break
        value /= divisor
        unit = next_unit

    if unit == "century" and value >= 1_000_000:
        return "millions of centuries"

    value = round(value, 1) if value < 10 else round(value)
    plural = "" if value == 1 else "s"
    return f"{_format_number(value)} {unit}{plural}"
